lib/Runtime/Library/JavascriptObject.cpp

```diff
@@ -858,10 +859,12 @@ namespace Js
 
         for (uint i = 0; i < ownPropsArray->GetLength(); i++)
         {
-            ownPropsArray->DirectGetItemAt(i, &propKey);
+            BOOL getPropResult = ownPropsArray->DirectGetItemAt(i, &propKey);
+            Assert(getPropResult);
```

Comment:
This opened up a good discussion in the office :)  @Yongqu @tcare and I got together and discussed the design of the APIs we have.

1. Currently the array you're calling `DirectGetItemAt` is guaranteed by our implementation to have items in order up to its length.  However we cannot convey this guarantee in the type system and so have no compile time guarantee.  In general this may not be true of JavascriptArray, elements can be deleted causing `DirectGetItemAt` to return `FALSE` for example.

 Because we cannot make a compile time guarantee here, we would like to see the return value handled in release builds.  Probably the best thing to do is Assert if it is false, and then skip the loop body and go to the next element in the array, since this shouldn't happen, but we don't want to use `propKey` if it wasn't guaranteed to be set to anything legitimate.

2. `JavascriptConversion::ToPropertyKey` should return `void`.  It is wasting time by having a confusing bool return type when it always succeeds.  This was poor API design but we can address it.  Not in this PR.  I'll open a separate issue.  I took a look and almost all callers of this API ignore the return value.  I think you should do the same and we'll change the return type to void in another change soon.

3. I verified that `JavascriptOperators::GetOwnPropertyKeys` always returns a `JavascriptArray*` regardless of which control path it takes.  `JavascriptOperators` APIs have traditionally been `Var` or `BOOL` returning, high level operations matching closely with the spec.  Returning `Var` makes sense particularly when the JIT leverages these APIs as helper calls.  However `GetOwnPropertyKeys` is used only internally to our runtime (not used by JIT) and all call sites immediately convert its result back to `JavascriptArray*`.  There's no reason we should conver to `Var` and back to `JavascriptArray*`, C++ is strongly typed, we should take advantage of that.

 I will open an issue to track making this improvement to our code base.  Let's leave it out of scope for your PR.  Since we're leaving it out though, your PR should be resilient, so you should have a code path that handles `JavascriptArray::Is(ownPropertyKeys)` returning false.  In that case you could just set ownPropertyKeys to a new zero length array via `scriptContext->GetLibrary()->CreateArray(0)`.